### PR TITLE
Finish implementing schema-less impl, switch all internal registrations to it

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #35058 C++ smoketests for the API.
* #34668 Python-driven testing for the new operator registration API
* #34620 Minor OperatorName cleanup
* #34618 Improved error checking
* #34601 Implement schema matching inside Dispatcher
* **#34456 Finish implementing schema-less impl, switch all internal registrations to it**
* #34455 Add torch::jit::parseName (mirroring parseSchema/parseSchemaOrName)
* #34221 Relax internal invariants of Dispatcher in preparation for schema-less registration

- 'operator_name_with_overload' is the schema string but minus the actual
  schema.  (Unfortunately, it's done in a very hacky way right now)
- Finish implementing register/deregistering schemas on DispatchKeyExtractor
  and DispatchTable
- Inline findOrRegisterSchema_ into its use site
- Delete alias analysis kind merging entirely.  Instead, use an impl()-only
  registration
- Fix bug in addRegistrationListener, where we should only eagerly register
  operators which actually have schemas (i.e., their def_count > 0)
- Add cloneWithName functionality to FunctionSchema
- Switch the new-style def()/impl() APIs to interact directly with Dispatcher,
  rather than indirecting through
- Unconditionally generate schema registration, even when type_method_dispatch
  is a dict

Notable TODOs:
- Add more unit tests
- Move inferred schema matching inside dispatcher (as in general, we will
  not have enough information to do the test at declaration site)
- Don't allow redefining operators with schema already, unless it's
  a legacy call
- Move other registrations to new style API (e.g., hand written registrations,
  and VariableType registrations)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20330414](https://our.internmc.facebook.com/intern/diff/D20330414)

### CODE REVIEW

File: aten/src/ATen/core/dispatch/Dispatcher.cpp [hunk 1]
```diff
@@ -112,14 +81,34 @@ RegistrationHandleRAII Dispatcher::registerDef(FunctionSchema schema) {
...
+      // Call the hook after the operator is setup!
+      listeners_->callOnOperatorRegistered(op);
+      return op;
+    } else {
+      // Re-registration for BC reasons, test for consistency
+      OperatorHandle op(*found);
+      TORCH_INTERNAL_ASSERT(op.operatorIterator_->def_and_impl_count >= 1);
+      // NB: the operator may have been impl'ed but not def'ed
+      if (op.operatorIterator_->def_count == 0) {
+        op.operatorIterator_->op.registerSchema(std::move(schema));
```
Comment (smessmer) [inline aten/src/ATen/core/dispatch/Dispatcher.cpp]: I think you have to call the listener here too, listener should trigger at the first def() call even if there's already an impl. Maybe it's better to keep the listener handling outside of this if/else, similar to how it was before. Also make sure to add a test case for this!

File: aten/src/ATen/core/dispatch/Dispatcher.cpp [hunk 2]
```diff
@@ -112,14 +81,34 @@ RegistrationHandleRAII Dispatcher::registerDef(FunctionSchema schema) {
...
+      operators_.emplace_back(std::move(schema));
+      OperatorHandle op(--operators_.end());
+      operatorLookupTable_.write([&] (ska::flat_hash_map<OperatorName, OperatorHandle>& operatorLookupTable) {
+        operatorLookupTable.emplace(op_name, op);
+      });
+      // Call the hook after the operator is setup!
+      listeners_->callOnOperatorRegistered(op);
+      return op;
+    } else {
+      // Re-registration for BC reasons, test for consistency
```
Comment (smessmer) [inline aten/src/ATen/core/dispatch/Dispatcher.cpp]: This is not only for bc reasons, right? It also happens when there was already an impl() but no def() yet and now a def() is coming in. Right? I think the TORCH_CHECK below is the only line that is there for bc reasons only.

File: aten/src/ATen/core/dispatch/Dispatcher.cpp [hunk 3]
```diff
@@ -112,14 +81,34 @@ RegistrationHandleRAII Dispatcher::registerDef(FunctionSchema schema) {
 
   OperatorName op_name = schema.operator_name();
 
-  auto op = findOrRegisterSchema_(std::move(schema));
+  const auto found = findSchema(op_name);
```
Comment (smessmer) [inline aten/src/ATen/core/dispatch/Dispatcher.cpp]: Maybe rename this to `findSchemaOrName` since it will also return operators that had an impl() but not a def() yet. Or maybe `findOperatorForName` or something like that?

File: aten/src/ATen/core/op_registration/op_registration.cpp [hunk 4]
```diff
@@ -153,27 +153,50 @@ namespace {
...
 
-Module&& Module::def(const char* schema) && {
-  register_.op(c10::RegisterOperators::options()
-    .schema(addNamespace(ns_, schema))
-    .aliasAnalysis(c10::AliasAnalysisKind::FROM_SCHEMA));
+// TODO: add overloads that take FunctionSchema directly
+Module&& Module::def(const char* schema_str) && {
+  auto schema = torch::jit::parseSchema(addNamespace(ns_, schema_str));
+  schema.setAliasAnalysis(c10::AliasAnalysisKind::FROM_SCHEMA); // TODO: clean this up
+  registrars_.emplace_back(Dispatcher::singleton().registerDef(std::move(schema)));
```
Comment (smessmer) [inline aten/src/ATen/core/op_registration/op_registration.cpp]: nit: can probably be `push_back`.

File: aten/src/ATen/core/op_registration/op_registration.cpp [hunk 5]
```diff
@@ -153,27 +153,50 @@ namespace {
...
-Module&& Module::impl(const char* name_or_schema, CppFunction&& f) && {
-  register_.op(c10::RegisterOperators::options()
-    .schema(addNamespace(ns_, name_or_schema))
-    // NB: Don't specify AliasAnalysis; the def() is expected to provide
-    // this
-    .kernel(f.dispatch_key_, std::move(f.func_), std::move(f.schema_)));
+Module&& Module::impl(const char* name_str, CppFunction&& f) && {
+  auto name = torch::jit::parseName(addNamespace(ns_, name_str));
+  // TODO: check that inferred schema matches real schema
+  registrars_.emplace_back(Dispatcher::singleton().registerImpl(name, f.dispatch_key_, std::move(f.func_)));
```
Comment (smessmer) [inline aten/src/ATen/core/op_registration/op_registration.cpp]: std::move(name)

File: aten/src/ATen/function_wrapper.py [hunk 6]
```diff
@@ -1203,16 +1205,16 @@ def gen_namespace_function(option, multidispatch_tensors):
         # we just implement it in the base Type.  This is exposed
         # in Declarations.yaml via a field named 'abstract'.
         abstract = False
+        op_registrations.append(OpRegistration(
```
Comment (smessmer) [inline aten/src/ATen/function_wrapper.py]: Why this change?

**Review by smessmer — CHANGES_REQUESTED** [review]
Generally looks good but I don't want to land this without a comprehensive suite of test cases. The API is now not a trivial wrapper of a well-tested API anymore. Feel free to add tests in a separate PR but please run them before you land this PR.

**Review by bhosmer — APPROVED** [review]
[Edit: oops, didn't actually mean to approve, given @smessmer's change requests. Definitely want to get to agreement on those.]

A few observations inline - apologies if the suggestions are based on misreadings, had to follow some stuff through the earlier PR and could easily have gotten things wrong.

Aside: something I kept wanting to explore as I read through this was getting schemas (and maybe even names) out of the operator entries entirely, since at first blush they only seem to be used as sentinels.

File: aten/src/ATen/core/dispatch/Dispatcher.cpp (hunk 1 above)
Comment (bhosmer) [reply aten/src/ATen/core/dispatch/Dispatcher.cpp]: @smessmer that's how I'm reading the code as well - we wind up here if `registerImpl` is called before `registerDef` on this op.

File: aten/src/ATen/core/dispatch/Dispatcher.cpp (hunk 2 above)
Comment (ezyang) [reply aten/src/ATen/core/dispatch/Dispatcher.cpp]: yes. I've fixed this already in a later patch

File: aten/src/ATen/core/dispatch/Dispatcher.cpp (hunk 1 above)
Comment (ezyang) [reply aten/src/ATen/core/dispatch/Dispatcher.cpp]: Yes this is a bug.

File: aten/src/ATen/core/dispatch/Dispatcher.cpp (hunk 3 above)
Comment (ezyang) [reply aten/src/ATen/core/dispatch/Dispatcher.cpp]: findOperatorForName seems better.

File: aten/src/ATen/core/op_registration/op_registration.cpp (hunk 4 above)
Comment (ezyang) [reply aten/src/ATen/core/op_registration/op_registration.cpp]: Isn't it a wash here? In both cases you're going to the call the move constructor to setup the entry in `registrars_`

File: aten/src/ATen/function_wrapper.py (hunk 6 above)
Comment (ezyang) [reply aten/src/ATen/function_wrapper.py]: I need to register schemas EVEN if there is a dictionary type method dispatch. Because those individual dictionary entries aren't going to have schemas associated with them.

Actually, I had to fix this code up some more in later PR, so the latest version is now:

```
        abstract = False
        if isinstance(type_method_dispatch, dict):
            abstract = True
            # Having manual_kernel_registration for an abstract method doesn't make sense.
            assert not option['manual_kernel_registration']
            op_registrations.append(OpRegistration(
                operator_name=OPERATOR_NAME.substitute(option),
                registration_code=SCHEMA_REGISTRATION.substitute(option)))
        else:
            top_env['type_method_declarations'].append(NATIVE_DISPATCH_DECLARATION.substitute(option))
            top_env['type_method_definitions'].append(NATIVE_DISPATCH_DEFINITION_DEFAULT.substitute(option))
            if not option['manual_kernel_registration']:
                op_registrations.append(OpRegistration(
                    operator_name=OPERATOR_NAME.substitute(option),
                    registration_code=SCHEMA_REGISTRATION.substitute(option)))
                if option['use_c10_dispatcher'] == 'full':
                    op_registrations.append(OpRegistration(
                        operator_name=OPERATOR_NAME.substitute(option),
                        registration_code=DEFAULT_FUNCTION_REGISTRATION.substitute(option)))
                else:
                    assert option['use_c10_dispatcher'] == 'unboxed_only'
                    op_registrations.append(OpRegistration(
                        operator_name=OPERATOR_NAME.substitute(option),
                        registration_code=DEFAULT_UNBOXEDONLY_FUNCTION_REGISTRATION.substitute(option)))
```

File: aten/src/ATen/core/dispatch/Dispatcher.cpp [hunk 7]
```diff
@@ -112,14 +81,34 @@ RegistrationHandleRAII Dispatcher::registerDef(FunctionSchema schema) {
...
+  OperatorHandle op = [&] {
+    if (found == c10::nullopt) {
+      // Initial def
+      operators_.emplace_back(std::move(schema));
+      OperatorHandle op(--operators_.end());
+      operatorLookupTable_.write([&] (ska::flat_hash_map<OperatorName, OperatorHandle>& operatorLookupTable) {
+        operatorLookupTable.emplace(op_name, op);
+      });
+      // Call the hook after the operator is setup!
+      listeners_->callOnOperatorRegistered(op);
```
[the remark this answers is not in the document]
Comment (ezyang) [reply aten/src/ATen/core/dispatch/Dispatcher.cpp]: def count is not public. What they do need setup, is that they need to be able to retrieve the schema from the operator handle when running the listener.

File: aten/src/ATen/core/dispatch/DispatchKeyExtractor.h [hunk 8]
```diff
@@ -105,6 +105,14 @@ struct CAFFE2_API DispatchKeyExtractor final {
     return DispatchKeyExtractor(0);
   }
 
+  void registerSchema(const FunctionSchema& schema) {
```
[the remark this answers is not in the document]
Comment (ezyang) [reply aten/src/ATen/core/dispatch/DispatchKeyExtractor.h]: 1. FunctionSchema still seems right, it lets Sebastian, e.g., decide to store a bitmap instead of arity, without having to plumb a constructor change as well in the patch.
2. I agree. But I don't want to add a guard inside the boxed calling hotpath. Maybe we can arrange for operator lookup to NOT return the operator if it doesn't have a schema.
3. No, `def_count` can be greater than 1, because legacy registrations repeatedly register schema.

File: aten/src/ATen/core/dispatch/Dispatcher.cpp [hunk 9]
```diff
@@ -112,14 +81,34 @@ RegistrationHandleRAII Dispatcher::registerDef(FunctionSchema schema) {
...
-  auto op = findOrRegisterSchema_(std::move(schema));
+  const auto found = findSchema(op_name);
+  OperatorHandle op = [&] {
+    if (found == c10::nullopt) {
+      // Initial def
+      operators_.emplace_back(std::move(schema));
+      OperatorHandle op(--operators_.end());
+      operatorLookupTable_.write([&] (ska::flat_hash_map<OperatorName, OperatorHandle>& operatorLookupTable) {
+        operatorLookupTable.emplace(op_name, op);
+      });
```
[the remark this answers is not in the document]
Comment (ezyang) [reply aten/src/ATen/core/dispatch/Dispatcher.cpp]: Yes this should work.

File: aten/src/ATen/core/dispatch/DispatchKeyExtractor.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (bhosmer) [reply aten/src/ATen/core/dispatch/DispatchKeyExtractor.h]: > 1. FunctionSchema still seems right, it lets Sebastian, e.g., decide to store a bitmap instead of arity, without having to plumb a constructor change as well in the patch.

Avoiding the constructor change works for me, but I'm not sure what you mean about deciding to store a bitmap vs arity - can you point me to where that's done? All I'm seeing is `DispatchTable` forwarding the schema to `DispatchKeyExtractor`, which just sets `num_args` off it then throws it away...

> 2. I agree. But I don't want to add a guard inside the boxed calling hotpath. Maybe we can arrange for operator lookup to NOT return the operator if it doesn't have a schema.

Agree, the guard should be "only ops in valid state should be addressable" rather than "make sure the op is in valid state when calling". 

That would also lend itself to a concurrency-friendly idiom where you load all the impls first, then commit the schema.

> 3. No, `def_count` can be greater than 1, because legacy registrations repeatedly register schema.

Ah I misread the code, despite your good comments :) are we committed to ongoing support for that behavior? (This is the LIFO-behavior-in-notebooks use case, right?)

[the remark this answers is not in the document]
Comment (smessmer) [reply aten/src/ATen/core/dispatch/DispatchKeyExtractor.h]: +1 on (2), i.e. only having valid ops being findable

File: aten/src/ATen/core/dispatch/Dispatcher.cpp (hunk 3 above)
Comment (ezyang) [reply aten/src/ATen/core/dispatch/Dispatcher.cpp]: This rename would be pretty invasive, as `findSchema` here is the literal top level findSchema that many clients use

```
  // ------------------------------------------------------------------------
  //
  // Accessing operators by schema
  //
  // ------------------------------------------------------------------------

  /**
   * Looks for an operator schema with the given name and overload name
   * and returns it if it is registered.
   * Returns nullopt otherwise.
   */
  c10::optional<OperatorHandle> findSchema(const OperatorName& operator_name);
```

I'm going to file an issue to do it later. https://github.com/pytorch/pytorch/issues/35038

[the remark this answers is not in the document]
Comment (ezyang) [reply aten/src/ATen/core/dispatch/Dispatcher.cpp]: issue filed to fix this later

File: aten/src/ATen/core/dispatch/DispatchKeyExtractor.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (ezyang) [reply aten/src/ATen/core/dispatch/DispatchKeyExtractor.h]: > Avoiding the constructor change works for me, but I'm not sure what you mean about deciding to store a bitmap vs arity - can you point me to where that's done? All I'm seeing is DispatchTable forwarding the schema to DispatchKeyExtractor, which just sets num_args off it then throws it away...

It's a hypothetical change, not in master, but one that has been done in PRs before. Here's a quote: https://github.com/pytorch/pytorch/pull/33313/files#diff-9ce4d20732d7b4c47e6a95c566f83ff6R101

> are we committed to ongoing support for that behavior? (This is the LIFO-behavior-in-notebooks use case, right?)

Not quite. People who were using the old RegisterOperators API always induce a "def" situation. My BC plan here is: (1) migrate all uses we have direct control over to the new API, (2) make it invalid to mix new-style with old-style registration, but preserve multiple defs for old api, (3) delete the old API entirely, at which point def can only happen once. I don't know how long it will take to get to (3) though.

The notebook stuff is for overriding impls. The plan on record here is to add a priority system and make notebook loading use priority to control ordering. I don't plan on doing this in the near future.

[the remark this answers is not in the document]
Comment (ezyang) [reply aten/src/ATen/core/dispatch/DispatchKeyExtractor.h]: Deprecation plan at https://github.com/pytorch/pytorch/issues/35040

[the remark this answers is not in the document]
Comment (ezyang) [reply aten/src/ATen/core/dispatch/DispatchKeyExtractor.h]: Jupyter plan at https://github.com/pytorch/pytorch/issues/35041

File: aten/src/ATen/core/dispatch/Dispatcher.cpp [hunk 10]
```diff
@@ -112,14 +81,34 @@ RegistrationHandleRAII Dispatcher::registerDef(FunctionSchema schema) {
 
   OperatorName op_name = schema.operator_name();
 
-  auto op = findOrRegisterSchema_(std::move(schema));
+  const auto found = findSchema(op_name);
+  OperatorHandle op = [&] {
```
[the remark this answers is not in the document]
Comment (ezyang) [reply aten/src/ATen/core/dispatch/Dispatcher.cpp]: @bhosmer Renaming here would go against the prevailing naming convention of the entire file, which means I'd have to do a rename of every occurrence of `op` for consistency. Is that really what you want?

File: aten/src/ATen/core/dispatch/DispatchKeyExtractor.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (bhosmer) [reply aten/src/ATen/core/dispatch/DispatchKeyExtractor.h]: > It's a hypothetical change, not in master, but one that has been done in PRs before. Here's a quote: [#33313 (files)](https://github.com/pytorch/pytorch/pull/33313/files#diff-9ce4d20732d7b4c47e6a95c566f83ff6R101)

Ah of course. I know about this change, just didn't make the connection that it would be happening here. Thanks!

Thanks also for the legacy API and Jupyter disambiguation.

File: aten/src/ATen/core/dispatch/Dispatcher.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (bhosmer) [reply aten/src/ATen/core/dispatch/Dispatcher.cpp]: Ahahaha I'm hearing it like, is that _really_ what you want? 😁 My impression was that `s/handle/op/` was a recent change of yours, but either way, I wouldn't insist. Maybe if I'm in here I'll see what it feels like to make the various indirections more explicit in the names

File: aten/src/ATen/core/dispatch/Dispatcher.cpp (hunk 1 above)
Comment (ezyang) [reply aten/src/ATen/core/dispatch/Dispatcher.cpp]: Fixed!

File: aten/src/ATen/core/dispatch/Dispatcher.cpp (hunk 3 above)
Comment (ezyang) [reply aten/src/ATen/core/dispatch/Dispatcher.cpp]: What I'm actually going to do, is findSchema will only return if there is actually a schema, and findOperatorForName will give you a thing always

File: aten/src/ATen/core/dispatch/DispatchKeyExtractor.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (ezyang) [reply aten/src/ATen/core/dispatch/DispatchKeyExtractor.h]: This is fixed now! findSchema only returns ops with schema. findOperatorByName returns everything.

File: aten/src/ATen/core/dispatch/Dispatcher.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (ezyang) [reply aten/src/ATen/core/dispatch/Dispatcher.cpp]: It's not. For example, on master, here's Dispatcher.cpp fragment:

```
std::pair<RegistrationHandleRAII, OperatorHandle> Dispatcher::registerSchema(FunctionSchema schema) {
  // we need a lock to avoid concurrent writes
  std::lock_guard<std::mutex> lock(mutex_);

  OperatorName op_name = schema.operator_name();

  auto op = findOrRegisterSchema_(std::move(schema));

  ++op.operatorIterator_->refcount;
  if (1 == op.operatorIterator_->refcount) {
    // note: call listeners *after* operator is added, i.e. dispatcher is already valid for new op
    listeners_->callOnOperatorRegistered(op);
  }

  return std::make_pair(RegistrationHandleRAII([this, op, op_name] {
    deregisterSchema_(op, op_name);
  }), op);
}
```

op here is an OperatorHandle.

File: aten/src/ATen/core/dispatch/Dispatcher.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (ezyang) [reply aten/src/ATen/core/dispatch/Dispatcher.cpp]: New code nicer!

```
  OperatorName op_name = schema.operator_name();
  auto op = findOrRegisterName_(op_name);
        
  if (op.operatorIterator_->def_count == 0) {
    // NB: registerSchema is not idempotent! Only do it once!
    op.operatorIterator_->op.registerSchema(std::move(schema));
    listeners_->callOnOperatorRegistered(op);
  } else {
    TORCH_CHECK(op.schema() == schema, "Tried to register multiple operators with the same name and the same overload name but different schemas: ", schema, " vs ", op.schema());
    if (schema.isDefaultAliasAnalysisKind()) {
      // If the *new* schema is the default alias analysis kind, for BC, we
      // will accept it.  If we don't accept it, most extensions that override
      // existing operators will stop working (as they generally did not
      // specify alias information).  Remove this BC smoothing ASAP, because
      // if the two incompatible registrations live in the same compilation
      // unit, the order their static initializers run is unspecified, which
      // means that you may nondeterministically fail the subsequent test.
    } else if (op.schema().isDefaultAliasAnalysisKind()) {
      // If you POST-FACTO specify a non-default alias analysis kind after
      // we already have a schema for a function, complain loudly about it
      // (because this new implementation doesn't support merging in this
      // way).
      TORCH_CHECK(op.schema().aliasAnalysis() == schema.aliasAnalysis(),
        "Tried to define the schema for ", toString(op_name),
        " multiple times without providing an explicit alias analysis kind at each registration site.  "
        "This was previously permitted, but is now not allowed.  You should either explicitly specify the "
        "correct alias analysis kind at each site [",
        toString(op.schema().isDefaultAliasAnalysisKind() ? schema.aliasAnalysis() : op.schema().aliasAnalysis()),  
        "], or use the new Module::impl() API, which permits you to omit the schema entirely when "
        "specifying further implementations of an operator");
    } else {
      TORCH_CHECK(op.schema().aliasAnalysis() == schema.aliasAnalysis(),
        "Tried to define the schema for ", toString(op_name), " with different alias analysis kinds: ",
        toString(op.schema().aliasAnalysis()), " vs ", toString(schema.aliasAnalysis()));
    }   
  }     
    
  // NB: do not increment the counts until AFTER error checking
  ++op.operatorIterator_->def_count;
  ++op.operatorIterator_->def_and_impl_count;
```

**Comment by ezyang** [comment]
All CR comments here addressed locally.

File: aten/src/ATen/core/dispatch/Dispatcher.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (bhosmer) [reply aten/src/ATen/core/dispatch/Dispatcher.cpp]: 😍😍😍

File: aten/src/ATen/core/dispatch/Dispatcher.cpp (hunk 3 above)
Comment (smessmer) [reply aten/src/ATen/core/dispatch/Dispatcher.cpp]: I think it generally makes sense to also rename the top level one, `findSchema` is not accurate anymore.

File: aten/src/ATen/core/op_registration/op_registration.cpp (hunk 4 above)
Comment (smessmer) [reply aten/src/ATen/core/op_registration/op_registration.cpp]: yes it probably is a wash but push_back is the simpler mechanism. It's a nit though, feel free to keep as is.

**Review by smessmer — APPROVED** [review]
